cpp/include/cuspatial/detail/cartesian_product_iterator.cuh

```diff
@@ -0,0 +1,227 @@
+ *  A0 * B2 | u5 v5 w5
+ *  A1 * B0 | x0 x1
+ *  A1 * B1 | x2 x3 x4
+ *  A1 * B2 | x5
+ *  A1 * B0 | y0 z0 y1 z1
+ *  A2 * B1 | y2 z2 y3 z3 y4 z4
+ *  A2 * B2 | y5 z5
+ *
+ * @tparam OffsetIteratorA
+ * @tparam OffsetIteratorB
+ * @param num_elements_a  number of elements from all groups in A
+ * @param num_elements_b  number of elements from all groups in B
+ * @param num_groups_a    number of groups in A
+ * @param num_groups_b    number of groups in B
+ * @param group_offsets_a offsets for each group in A, plus an "end" offset
+ * @param group_offsets_b offsets for each group in B, plus an "end" offset
+ * @return auto           Thrust iterator of `cartesian_product_group_index`s. device-only.
+ *
+ * @note Can be used in conjunction with `make_transform_iterator` to create a single kernel which
+ * simultaneously computes distance metrics for multiple spaces in O(n^2) time using a single call
+ * to `reduce_by_key`.
+ */
+template <typename OffsetIteratorA, typename OffsetIteratorB>
+auto make_grouped_cartesian_product_iterator(int32_t const num_elements_a,
+                                             int32_t const num_elements_b,
+                                             int32_t const num_groups_a,
+                                             int32_t const num_groups_b,
+                                             OffsetIteratorA const group_offsets_a,
+                                             OffsetIteratorB const group_offsets_b)
+{
+  auto count_iter = thrust::make_counting_iterator(0);
+
+  using SizeFunctorLHS = detail::size_from_offsets_functor<OffsetIteratorA>;
+  using SizeFunctorRHS = detail::size_from_offsets_functor<OffsetIteratorB>;
+
+  auto group_sizes_a = thrust::make_transform_iterator(
+    count_iter, SizeFunctorLHS{num_groups_a, num_elements_a, group_offsets_a});
+  auto group_sizes_b = thrust::make_transform_iterator(
+    count_iter, SizeFunctorRHS{num_groups_b, num_elements_b, group_offsets_b});
+  auto group_lookup_a = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorA>{group_offsets_a, num_groups_a});
+  auto group_lookup_b = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorB>{group_offsets_b, num_groups_b});
+
+  using SizeIteratorLHS   = decltype(group_sizes_a);
+  using SizeIteratorRHS   = decltype(group_sizes_b);
+  using LookupIteratorLHS = decltype(group_lookup_a);
+  using LookupIteratorRHS = decltype(group_lookup_b);
+
+  using TraversalFunctor = cartesian_product_functor<OffsetIteratorA,
+                                                     OffsetIteratorB,
+                                                     SizeIteratorLHS,
+                                                     SizeIteratorRHS,
+                                                     LookupIteratorLHS,
+                                                     LookupIteratorRHS>;
+
+  auto traversal = TraversalFunctor{num_elements_b,
```

Comment:
In fact, all of the types are deducible from this call. You don't need the `using TraversalFunctor` at all, right?

```suggestion
  auto traversal = cartesian_product_functor(num_elements_b,
```

Reply:
That's likely. I'll give it a try.

Reply:
Hmm. I wasn't able to get this to work. I tried writing a constructor, but that did not make a difference. In fact, it just added more cruft. This function's primary purpose is to handle the type inference for _downstream_ construction, though, so these explicit types don't seem out of place to me. :man_shrugging: